File: implementations/pytorch_conv/benchmark.py
import yaml
import sys
from discretize_pde import discretize_heat_equation
import torch
import time
import numpy as np
filename = sys.argv[1]
with open(filename) as stream:
    node = yaml.safe_load(stream)
    N = node["discretization"]["N"]
    dt = node["discretization"]["dt"]
    t_end = node["t_end"]
    write_every = node["write_every"]
    device = "cuda" if torch.cuda.is_available() else "cpu"
    #benchmark
    num_iterations=5
    times=[]
    for i in range(num_iterations):
        start=time.time()
        discretize_heat_equation(N, dt, t_end, write_every,device=device)
        end=time.time()
        times.append((end-start))
    output_file="benchmark_results.yaml"
    yaml.dump({"time":{
            "mean":np.mean(times).item(),
            "std":np.std(times).item(),
            "min":np.min(times).item(),
            "max":np.max(times).item(),
            "median":np.median(times).item()
        },"time_unit":"s"},open(output_file,"w"))
    
    

File: implementations/julia_cuda_array/benchmark.jl
using YAML
using BenchmarkTools
include("discretize_pde.jl")
rm("output", recursive=true, force=true)
mkpath("output")
config_path = length(ARGS) == 1 ? ARGS[1] : "../../config.yaml"
config = YAML.load_file(config_path)
#display(config)
N = config["discretization"]["N"]
dt = config["discretization"]["dt"]
t_end = config["t_end"]
write_every = config["write_every"]

t = @benchmark discretize_heat_equation(N, Float32(dt), Float32(t_end), write_every)
display(t)
times= t.times ./ 1e9
YAML.write_file("benchmark_results.yaml", Dict(
    "time" => Dict(
        "mean" => mean(times),
        "std" => std(times),
        "median" => median(times),
        "min" => minimum(times),
        "max" => maximum(times)
    ),
    "time_unit" => "s",
    "memory" => t.memory
))